utils/src/main/java/software/amazon/awssdk/utils/async/AsyncBufferingSubscriber.java

```diff
@@ -0,0 +1,202 @@
+        subscription.request(1);
+    }
+
+    @Override
+    public void onNext(T item) {
+        if (item == null) {
+            subscription.cancel();
+            throw new NullPointerException("Item must not be null");
+        }
+
+        try {
+            buffer.add(new DataEvent<>(item));
+            flushBuffer();
+        } catch (Exception e) {
+            isStreamingDone = true;
+            subscription.cancel();
+            future.completeExceptionally(e);
+        }
+    }
+
+    private void deliverItem(T item) {
+        log.debug(() -> "Delivering next item, numRequestInFlight=" + numRequestsInFlight);
+        numRequestsInFlight.incrementAndGet();
+        consumer.apply(item).whenComplete((r, t) -> {
+            numRequestsInFlight.decrementAndGet();
+            if (!isStreamingDone) {
+                subscription.request(1);
+            } else {
+                flushBuffer();
+            }
+        });
+    }
+
+    private void flushBuffer() {
+        if (buffer.isEmpty()) {
+            if (isStreamingDone && numRequestsInFlight.get() == 0) {
+                future.complete(null);
+            } else {
+                subscription.request(1);
+            }
+            return;
+        }
+
+        if (isDelivering.compareAndSet(false, true)) {
+            Event<T> firstEvent = buffer.peek();
+            if (firstEvent != null && isCompleteEvent(firstEvent)) {
+                Event<T> event = buffer.poll();
+                handleCompleteEvent(event);
+                isDelivering.set(false);
+                return;
+            }
+
+            while (!buffer.isEmpty() && numRequestsInFlight.get() < maxConcurrentRequests) {
+                Event<T> event = buffer.poll();
+                if (event == null) {
+                    break;
+                }
+
+                if (isCompleteEvent(event)) {
+                    handleCompleteEvent(event);
+                    isDelivering.set(false);
+                    return;
+                }
+
+                DataEvent<T> dataEvent = (DataEvent<T>) event;
+                deliverItem(dataEvent.data);
+            }
+            isDelivering.set(false);
+        }
+    }
+
+    private void handleCompleteEvent(Event<T> event) {
+        isStreamingDone = true;
+        if (numRequestsInFlight.get() == 0) {
+            future.complete(null);
+        }
+    }
+
+    @Override
+    public void onError(Throwable t) {
+        handleError(t);
+    }
+
+    private void handleError(Throwable t) {
+        future.completeExceptionally(t);
+        buffer.clear();
+    }
+
+    @Override
+    public void onComplete() {
+        buffer.add(completeEvent);
+        flushBuffer();
+    }
+
+    private boolean isCompleteEvent(Event<T> event) {
+        return event.type() == EventType.COMPLETE;
+    }
+
+    private enum EventType {
+        DATA,
+        COMPLETE
+    }
+    
+    private interface Event<T> {
+        EventType type();
+    }
```

Comment:
I think this approach is fine (and pretty clean), but we could also consider avoiding the wrapper entirely by using a `Queue<Object>`, and declaring a `private static final COMPLETE_EVENT = new Object()`. If the direct reference comparison fails, we know the object is a `T`.